### ISSUE TITLE
Enable MPI from DockerSpawned Jupyter

### ISSUE THREAD

**Issue by robnagler** [jupyterhub/jupyterhub#587]
I am probably missing something about Jupyter/Hub so I apologize in advance for the rather long explanation and suggestion.

We'd like to be able to start MPI from a Jupyter container, which was started by JupyterHub. I've proven that it's possible, but it required quite a few workarounds due to how JupyterHub acquires the IP and port for the Jupyter container.

In order to enable MPI, the container must be started with network_mode=host (--net=host). MPI requires an arbitrary set of ports be open. Both the slaves and the master listen on TCP sockets.

When network_mode=host is set, the Jupyter container is sharing ports with all other processes on the host. Therefore, the Jupyter container needs to select a random port to use. The way this works with LocalSpawner is that it opens a random port, closes it, and then passes that port number to the Jupyter process. There's a race condition with this algorithm. With DockerSpawner and network_mode=host, this algorithm can't work, because the spawned containers may not be on the local machine and/or the DockerSpawner may not be running in a network-privileged container.

Docker's inspect API returns no NetworkSettings when the container is using the host's network. Docker simply has not control over the network so it doesn't know what to report. This is not necessarily a problem except that DockerSpawner has to return an ip and port to JupyterHub, and DockerSpawner.get_ip_and_port expects the Docker API to report the network configuration.

One workaround is to use Docker's exec API to talk to the container and get it to dump its network settings. For a simple network configuration, this isn't a problem: you can interrogate /proc/net/tcp. However, for a complex network, that's not so easy. You don't really know which interface will be used to connect the JupyterHub/Proxy to Jupyter -- even then, technically, the routes might not be symmetric, but we can skip this case. :)

As a relevant aside, I have noticed that when a Jupyter container has problems starting, it may be "up" for long enough for JupyterHub to connect to the container's proxy before Jupyter dies. This is another race condition, because JuypterHub has to poll Jupyter instead of Jupyter notifying JupyterHub all is ok until it polls to see if the container is ok.

My suggestion is to have Jupyter notify JupyterHub that it's ok using the Hub API. When that notification happens, JupyterHub will know that Jupyter is ready to serve requests and from which IP/port Jupyter is listening on. This would eliminate the two race conditions as well as replace the polling with a simple timer to handle the case Jupyter doesn't come up. Jupyter would know that it needed to callback to the Hub via an environment variable.

Does this seem like a reasonable proposal?

**Comment by minrk** [comment]
> Does this seem like a reasonable proposal?

Yes, absolutely. I proposed something similar to this in #302, but haven't gotten around to it. It will be useful to take DockerSpawner as a use case when trying this out, because there are two distinct cases:

1. host networking, where the best option is for the SU server to pick its port and notify
2. docker networking, where the SU server *can't* know its own port information wrt the Hub, and the current logic is preferable.

As an aside, are you exposing host MPI, or MPI across docker containers? If it is across docker containers, could overlay networks avoid the need for host networking?

**Comment by robnagler** [comment]
Thanks, @minrk. I'm glad I'm not the only one to run into this.

I think there are two problems being solved with one callback:

A. The server is running ok
B. Here's the connection

When the Spawner starts the container, it doesn't have to pass a port (both in the random_port and net=host cases). Instead it reports back to the Hub None for port and/or IP. The Spawner can be considered authoritative when it does report back the port and/or IP. Until the (A) and (B) state is received, the Hub doesn't need to do anything with the IP and port. It's only after the callback that the proxy needs to be setup, and the Hub can then elect which IP:port to use.

> As an aside, are you exposing host MPI, or MPI across docker containers? 

Across containers.

> If it is across docker containers, could overlay networks avoid the need for host networking?

For a simple MPI test, I think so. From my perspective, it seems like more work than net=host. For a "real" MPI cluster, I'm not sure an overlay network solves the problem. I think it might be a problem from the perspective of network selection. MPI chooses the fastest network between hosts. With overlay networks, I'm guessing the hardware is hidden from MPI. I'm also not sure yet what level of privilege is required for network selection. All of this tech is new to me so your input would be very helpful.

What I'm hoping to do is run the MPI cluster as an ordinary user. The cluster works fine this way, and the SU network configuration was the missing piece. After a night's sleep, running MPI directly from the container is not ideal due to queue management issues. However, using Grid Engine or something similar seems to create more problems than it solves for our particular case. 

Way off topic: Do you know of a dynamically configured cluster queue manager that allows user selectable containers? That's the ultimate goal. Star Cluster might work, but even it does not seem to support user-selectable containers out of the box. Another part of this project is testing particle beam simulation codes in the cloud vs on bare metal. Feel free to DM if you'd like to help. :)

BTW, you can see the configuration evolving in [radiasoft/salt-conf](https://github.com/radiasoft/salt-conf).

**Comment by willingc** [comment]
@robnagler @minrk Should this be kept open?

**Comment by minrk** [comment]
I never did answer the last question about queues. I don't know of one, though. The [everware](http://everware.xyz) folks are providing user-selectable images based on GitHub repos (like binder for JupyterHub). #302 already covers the main feature proposal of switching to a web hook for registration.

**Comment by robnagler** [comment]
I'm fine closing this. I'll create a new issue when we are ready to do
something about this.

**Comment by willingc** [comment]
Thanks @minrk and @robnagler :sunny:
